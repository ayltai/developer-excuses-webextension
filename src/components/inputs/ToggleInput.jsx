import { Box, FormControlLabel, makeStyles, Switch, } from '@material-ui/core';
import PropTypes from 'prop-types';
import React from 'react';

export const ToggleInput = props => {
    const classes = makeStyles(theme => ({
        margin : {
            marginBottom : theme.spacing(2),
        },
        label  : {
            marginLeft : 0,
        },
    }))();

    const [ checked, setChecked, ] = React.useState(props.checked || false);

    return (
        <Box className={classes.margin}>
            <FormControlLabel
                className={classes.label}
                control={
                    <Switch
                        color='default'
                        checked={checked}
                        onChange={() => {
                            const value = !checked;

                            setChecked(value);

                            if (props.onChange) props.onChange(value);
                        }} />
                }
                label={props.label}
                labelPlacement='start' />
        </Box>
    );
};

ToggleInput.propTypes = {
    label    : PropTypes.string,
    checked  : PropTypes.bool,
    onChange : PropTypes.func,
};
