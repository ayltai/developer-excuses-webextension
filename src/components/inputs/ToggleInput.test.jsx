import { FormControlLabel, } from '@material-ui/core';
import { mount, shallow, } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import { ToggleInput, } from './ToggleInput';

describe('<ToggleInput />', () => {
    it('renders correctly', () => {
        expect(renderer.create(
            <ToggleInput
                label='dummy'
                checked={false}
                onChange={() => {}} />
        ).toJSON()).toMatchSnapshot();
    });

    it('mounts without error', () => {
        mount(
            <ToggleInput
                label='dummy'
                checked={false}
                onChange={() => {}} />
        );
    });

    it('triggers onChange when toggled', () => {
        const onChange = jest.fn();

        shallow(
            <ToggleInput
                label='dummy'
                checked={false}
                onChange={onChange} />)
            .find(FormControlLabel)
            .prop('control')
            .props
            .onChange();

        expect(onChange.mock.calls.length).toEqual(1);
    });

    it('triggers onChange when toggled', () => {
        let value = false;

        const onChange = newValue => {
            value = newValue;
        };

        shallow(
            <ToggleInput
                label='dummy'
                checked={value}
                onChange={onChange} />)
            .find(FormControlLabel)
            .prop('control')
            .props
            .onChange();

        expect(value).toBeTruthy();
    });

    it('does not trigger onChange when toggled', () => {
        let value = false;

        shallow(
            <ToggleInput
                label='dummy'
                checked={value} />)
            .find(FormControlLabel)
            .prop('control')
            .props
            .onChange();

        expect(value).toBeFalsy();
    });
});
