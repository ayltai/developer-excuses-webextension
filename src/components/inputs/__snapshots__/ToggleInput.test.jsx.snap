// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleInput /> renders correctly 1`] = `
<div
  className="MuiBox-root MuiBox-root-3 makeStyles-margin-1"
>
  <label
    className="MuiFormControlLabel-root makeStyles-label-2 MuiFormControlLabel-labelPlacementStart"
  >
    <span
      className="MuiSwitch-root"
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-4 MuiSwitch-switchBase"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            checked={false}
            className="PrivateSwitchBase-input-7 MuiSwitch-input"
            onChange={[Function]}
            type="checkbox"
          />
          <span
            className="MuiSwitch-thumb"
          />
        </span>
      </span>
      <span
        className="MuiSwitch-track"
      />
    </span>
    <span
      className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      dummy
    </span>
  </label>
</div>
`;
